chall03/migferna.py

```diff
@@ -0,0 +1,54 @@
+#!/usr/bin/env python3
+
+import sys
+
+def snail(matrix):
+    res = []
+    if len(matrix) == 0:
+        return res
+    row_begin = 0
+    row_end = len(matrix) - 1
+    col_begin = 0
+    col_end = len(matrix[0]) - 1
+
+    while row_begin <= row_end and col_begin <= col_end:
+        for i in range(col_begin, col_end + 1):
+            res.append(matrix[row_begin][i])
+        row_begin += 1
+
+        for i in range(row_begin, row_end+1):
+            res.append(matrix[i][col_end])
+        col_end -= 1
+
+        if row_begin <= row_end:
+            for i in range(col_end, col_begin - 1, -1):
+                res.append(matrix[row_end][i])
+        row_end -= 1
+
+        if (col_begin <= col_end):
+            for i in range(row_end, row_begin - 1, -1):
+                res.append(matrix[i][col_begin])
+        col_begin += 1
+    return res
```

Comment:
This part works fine, but it looks like C code, there are a lot of python features you could have taken advantage of.

First of all you just want to wait that your arg list is empty, knowing that `not []` returns True, you could have written.
```diff
-    row_begin = 0
-    row_end = len(matrix) - 1
-    col_begin = 0
-    col_end = len(matrix[0]) - 1
-
-    while row_begin <= row_end and col_begin <= col_end:
+    while matrix:
```

You could have taken advantage of this on the whole thing, but to do so, as strings are not editable in python, you would have had to convert all the rows into arrays. A simple solution would be `array = [char for char in string]`.

Later you could have taken advantage of the `.pop` method that returns an element of the array at the given index by removing it from the said array.  
Also you could have used the following array operator `[::-1]` that returns a reversed copy of an array (or a string)...

I'll let you read the solution when it's uploaded for more details :)